src/dtx/dtx_common.c

```diff
@@ -784,56 +775,40 @@ dtx_batched_commit_deregister(struct ds_cont_child *cont)
 	int				 rc;
 
 	D_ASSERT(cont != NULL);
-	if (cont->sc_closing) {
-		D_ASSERT(cont->sc_dtx_flush_cbdata != NULL);
-
-		future = cont->sc_dtx_flush_cbdata;
-		goto wait;
-	}
+	D_ASSERT(cont->sc_open == 0);
 
 	head = &dss_get_module_info()->dmi_dtx_batched_list;
 	d_list_for_each_entry(dbca, head, dbca_link) {
 		if (uuid_compare(dbca->dbca_cont->sc_uuid,
 				 cont->sc_uuid) != 0)
 			continue;
 
-		if (--(dbca->dbca_shares) > 0)
-			return;
-
-		/* Notify the dtx_batched_commit ULT to flush the
-		 * committable DTXs via setting @sc_closing as 1.
+		/*
+		 * Notify the dtx_batched_commit ULT to flush the
+		 * committable DTXs.
 		 *
 		 * Then current ULT will wait here until the DTXs
 		 * have been committed by dtx_batched_commit ULT
 		 * that will wakeup current ULT.
 		 */
-		D_ASSERT(cont->sc_dtx_flush_cbdata == NULL);
-		D_ASSERT(cont->sc_dtx_flush_wait_count == 0);
-
+		D_ASSERT(dbca->dbca_deregistering == NULL);
 		rc = ABT_future_create(1, NULL, &future);
-		cont->sc_closing = 1;
 		if (rc != ABT_SUCCESS) {
 			D_ERROR("ABT_future_create failed for DTX flush on "
 				DF_UUID" %d\n", DP_UUID(cont->sc_uuid), rc);
 			return;
 		}
 
-		cont->sc_dtx_flush_cbdata = future;
-		goto wait;
-	}
-
-	D_ASSERT(0);
-
-wait:
-	cont->sc_dtx_flush_wait_count++;
-	rc = ABT_future_wait(future);
-	D_ASSERTF(rc == ABT_SUCCESS, "ABT_future_wait failed "
-		  "for DTX flush (2) on "DF_UUID": rc = %d\n",
-		  DP_UUID(cont->sc_uuid), rc);
+		dbca->dbca_deregistering = future;
+		rc = ABT_future_wait(future);
```

Comment:
What will happen if the close is resent because the former close is blocked here?

Reply:
Indeed, we should call ds_cont_local_close() before deregister in cont_close_hdl().